### ISSUE TITLE
pilot's discovery becomes not ready 

### ISSUE THREAD

**Issue by sbezverk** [istio/istio#12308]
After working for 20 something minutes pilot got into non ready state:

```
 Normal   Scheduled    42m                   default-scheduler                   Successfully assigned istio-system/istio-pilot-755579f496-rcv4d to kube-7.sbezverk.cisco.com
  Warning  FailedMount  42m                   kubelet, kube-7.sbezverk.cisco.com  MountVolume.SetUp failed for volume "istio-pilot-service-account-token-zdvg5" : couldn't propagate object cache: timed out waiting for the condition
  Warning  FailedMount  42m                   kubelet, kube-7.sbezverk.cisco.com  MountVolume.SetUp failed for volume "istio-certs" : couldn't propagate object cache: timed out waiting for the condition
  Normal   Pulling      42m                   kubelet, kube-7.sbezverk.cisco.com  pulling image "192.168.80.240:4000/istio/pilot:v1.2.1_ip6tablesv6_local_cluster"
  Normal   Pulled       42m                   kubelet, kube-7.sbezverk.cisco.com  Successfully pulled image "192.168.80.240:4000/istio/pilot:v1.2.1_ip6tablesv6_local_cluster"
  Normal   Created      42m                   kubelet, kube-7.sbezverk.cisco.com  Created container
  Normal   Pulling      42m                   kubelet, kube-7.sbezverk.cisco.com  pulling image "192.168.80.240:4000/istio/proxyv2:v1.2.1_ip6tablesv6_local_cluster"
  Normal   Started      42m                   kubelet, kube-7.sbezverk.cisco.com  Started container
  Normal   Pulled       41m                   kubelet, kube-7.sbezverk.cisco.com  Successfully pulled image "192.168.80.240:4000/istio/proxyv2:v1.2.1_ip6tablesv6_local_cluster"
  Normal   Created      41m                   kubelet, kube-7.sbezverk.cisco.com  Created container
  Normal   Started      41m                   kubelet, kube-7.sbezverk.cisco.com  Started container
  Warning  Unhealthy    41m                   kubelet, kube-7.sbezverk.cisco.com  Readiness probe failed: Get http://[2001:470:b16e:1001::5678]:8080/ready: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
  Warning  Unhealthy    15s (x19 over 9m15s)  kubelet, kube-7.sbezverk.cisco.com  Readiness probe failed: Get http://[2001:470:b16e:1001::5678]:8080/ready: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting head
```

**Comment by sbezverk** [comment]
See complete logs collected from both containers:

https://gist.github.com/sbezverk/6abcb6885ca3469680eceb3c48cd3ed1

**Comment by sbezverk** [comment]
@costinm @rshriram  I left pilot in the broken state, let me know if you need me to collect additional things..

**Comment by sbezverk** [comment]
Adding Galley logs:

https://gist.github.com/sbezverk/b42c908ecb859b00ccc972e9ac1bb6f1

**Comment by duderino** [comment]
Making this P0 until proven innocent

**Comment by rshriram** [comment]
logs show pilot not ready.. not the proxy

**Comment by sbezverk** [comment]
[pprof.log](https://github.com/istio/istio/files/2943403/pprof.log)

Please check if it is a valid file.

**Comment by andraxylia** [comment]
I see some issues with the certificates in the pilot log:
2019-03-06 21:51:13.711][56][info][main] [external/envoy/source/server/server.cc:512] exiting
Invalid path: /etc/certs/cert-chain.pem
2019-03-06T21:51:13.712419Z     warn    Epoch 0 terminated with an error: exit status 1
2019-03-06T21:51:13.712441Z     warn    Aborted all epochs
2019-03-06T21:51:13.712467Z     warn    Failed to delete config file /etc/istio/proxy/envoy-rev0.json for 0, remove /etc/istio/proxy/envoy-rev0.json: no such file or directory
2019-03-06T21:51:13.712477Z     info    Epoch 0: set retry delay to 25.6s, budget to 2
2019-03-06T21:51:39.312639Z     info    Reconciling retry (budget 2)
2019-03-06T21:51:39.312713Z     info    Epoch 0 starting

**Comment by andraxylia** [comment]
The proxy does have issues and it keeps restarting:
[2019-03-06 21:50:48.229][25][debug][misc] [external/envoy/source/common/filesystem/filesystem_impl.cc:105] Unable to determine canonical path for /etc/certs/cert-chain.pem: No such file or directory
[2019-03-06 21:50:48.229][25][critical][main] [external/envoy/source/server/server.cc:86] error initializing configuration '/etc/istio/proxy/envoy.yaml': Invalid path: /etc/certs/cert-chain.pem

@sbezverk can you manually inspect if the certificate is there?

**Comment by sbezverk** [comment]
@andraxylia yep, it is there now:
```
root@istio-pilot-755579f496-rcv4d:/# ls /etc/certs/cert-chain.pem
/etc/certs/cert-chain.pem
root@istio-pilot-755579f496-rcv4d:/# 
```

**Comment by andraxylia** [comment]
This has nothing to do with MCP not being ready, it prints a message when returning false, unless there is some deadlock or delay longer than 5 secs in getting that mutex (there is a timeout of 5s in the readiness):

https://github.com/istio/istio/blob/e294254e6b0415f798e8f5fb02e6a1d4ca304c5f/pilot/pkg/config/coredatamodel/controller.go#L207

First time it printed it was:
21:51:25.785603Z     info    Configuration not synced

Last time this was printed was at:
21:51:25.785603Z     info    Configuration not synced

Pilot kept pushing 41 more minutes, last push at 22:26
2019-03-06T22:26:50.436236Z     info    ads     Push finished: 10.248859ms 

If there is some deadlock the trigger must have happened in the last 10 secs or so.

**Comment by ozevren** [comment]
TL;DR is, I don’t think this is instigated by Galley or MCP. The MCP implementation (or the related resource ready check) doesn’t seem to be a culprit either.

It is worth noting that when the readiness endpoint is curled within the pod it works. 


Here is the detailed analysis:

Galley gets stable Mixer clients, and unstable Pilot clients:
https://gist.github.com/sbezverk/b42c908ecb859b00ccc972e9ac1bb6f1#file-galley-logs-L444
Mixer 1:
```
2019-03-06T21:51:23.189877Z     info    mcp     MCP: connection {addr=[2001:470:b16e:1001::5677]:43302 id=1}: NEW (ResourceSource), supported collections: ...
2019-03-06T21:51:23.189987Z     info    mcp     MCP: connection {addr=[2001:470:b16e:1001::5677]:43302 id=1}: inc=false WATCH for istio/policy/v1beta1/instances
```

Pilot 1:
https://gist.github.com/sbezverk/b42c908ecb859b00ccc972e9ac1bb6f1#file-galley-logs-L631
```
2019-03-06T21:51:25.877416Z     info    mcp     MCP: connection {addr=[2001:470:b16e:1001::5678]:58204 id=2}: NEW (ResourceSource), supported collections: ...
2019-03-06T21:51:25.877478Z     info    mcp     MCP: connection {addr=[2001:470:b16e:1001::5678]:58204 id=2}: inc=false WATCH for istio/networking/v1alpha3/virtualservices
```

Mixer 2:
https://gist.github.com/sbezverk/b42c908ecb859b00ccc972e9ac1bb6f1#file-galley-logs-L680
```
2019-03-06T21:51:29.332552Z     info    mcp     MCP: connection {addr=[2001:470:b16e:1001::5676]:40196 id=3}: NEW (ResourceSource), supported collections: ...
2019-03-06T21:51:29.332611Z     info    mcp     MCP: connection {addr=[2001:470:b16e:1001::5676]:40196 id=3}: inc=false WATCH for istio/config/v1alpha2/legacy/redisquotas
```


Pilot starts with “not synced state”, and goes in this mode for some time:
https://gist.github.com/sbezverk/6abcb6885ca3469680eceb3c48cd3ed1#file-gistfile1-txt-L827
```
2019-03-06T21:50:45.285598Z     info    Configuration not synced: first push for [istio/networking/v1alpha3/destinationrules istio/authentication/v1alpha1/policies ...] not received
```

Eventually syncs:
https://gist.github.com/sbezverk/6abcb6885ca3469680eceb3c48cd3ed1#file-gistfile1-txt-L836
```
2019-03-06T21:51:25.785603Z     info    Configuration not synced: first push for [istio/networking/v1alpha3/envoyfilters istio/networking/v1alpha3/sidecars...] not received
2019-03-06T21:51:25.876682Z     info    mcp     (re)trying to establish new MCP sink stream
```

At this point Pilot is fully synced, and starts doling out config and does not complain about not being synced. This is a latch, meaning once HasSynced signals, it doesn’t seem to go back to its unsynced state:
https://github.com/istio/istio/blob/e294254e6b0415f798e8f5fb02e6a1d4ca304c5f/pilot/pkg/config/coredatamodel/controller.go#L207


After a while, Pilot 1's connection gets terminated by the gRPC stack due to too many pings.

From Galley:
https://gist.github.com/sbezverk/b42c908ecb859b00ccc972e9ac1bb6f1#file-galley-logs-L792
```
2019-03-06T21:53:44.880354Z     info    transport: Got too many pings from the client, closing the connection.
2019-03-06T21:53:44.880482Z     info    transport: loopyWriter.run returning. Err: transport: Connection closing
2019-03-06T21:53:44.880602Z     info    transport: http2Server.HandleStreams failed to read frame: read tcp [2001:470:b16e:1001::567c]:9901->[2001:470:b16e:1001::5678]:58204: use of closed network connection
2019-03-06T21:53:44.880708Z     info    mcp     MCP: connection {addr=[2001:470:b16e:1001::5678]:58204 id=2}: CLOSED
2019-03-06T21:53:44.880916Z     info    mcp     MCP: connection {addr=[2001:470:b16e:1001::5678]:58204 id=2}: TERMINATED "rpc error: code = Canceled desc = context canceled"
```

Corresponding log from Pilot:
https://gist.github.com/sbezverk/6abcb6885ca3469680eceb3c48cd3ed1#file-gistfile1-txt-L1011
```
2019-03-06T21:53:17.800712Z     info    ads     ADS: "[2001:470:b16e:1001::5675]:50064" router~2001:470:b16e:1001::5675~istio-ingressgateway-dddb87f65-bsf7k.istio-system~istio-system.svc.k7.sbezverk.cisco.com-1 terminated rpc error: code = Canceled desc = context canceled
2019-03-06T21:53:17.800761Z     info    transport: loopyWriter.run returning. connection error: desc = "transport is closing"
2019-03-06T21:53:44.880505Z     info    Client received GoAway with http2.ErrCodeEnhanceYourCalm.
2019-03-06T21:53:44.880750Z     error   mcp     Error receiving MCP resource: rpc error: code = Unavailable desc = transport is closing
2019-03-06T21:53:44.880799Z     error   mcp     Error receiving MCP response: rpc error: code = Unavailable desc = transport is closing
```

This dance goes on for some time, with some time gaps.

https://gist.github.com/sbezverk/b42c908ecb859b00ccc972e9ac1bb6f1#file-galley-logs-L845
```
2019-03-06T21:53:45.885142Z     info    mcp     Watch(): created watch 143 for istio/rbac/v1alpha1/rbacconfigs from group "default", version "0"
2019-03-06T21:58:04.884390Z     info    transport: Got too many pings from the client, closing the connection.
2019-03-06T21:58:04.884390Z     info    transport: Got too many pings from the client, closing the connection.
```

```
2019-03-06T21:58:05.887214Z     info    mcp     Watch(): created watch 159 for istio/networking/v1alpha3/gateways from group "default", version "0"
2019-03-06T22:06:24.889940Z     info    transport: Got too many pings from the client, closing the connection.
2019-03-06T22:06:24.889940Z     info    transport: Got too many pings from the client, closing the connection.
2019-03-06T22:06:24.890089Z     info    transport: loopyWriter.run returning. Err: transport: Connection closing
```

**Comment by andraxylia** [comment]
To follow up on my previous comment, there is no deadlock either in HasSynced() for MCP, so the readiness problem could come from the  KubeController.HasSynced().

**Comment by andraxylia** [comment]
@sbezverk does it work to get a running core dump of pit process with gcore https://rakyll.org/coredumps/ ?

**Comment by andraxylia** [comment]
@sbezverk could you please run kubectl get pod -oyaml  on the pilot pod and get the state of your containers, to see which 1/2 is actually down? The problem may be with the istio-proxy container not with pilot-discovery anymore. Maybe at some point it may have been not ready, but now curl works. On the other hand, I saw issues with certificates in istio-proxy in the logs and logs never shows proxy as recovered.

**Comment by sbezverk** [comment]
@andraxylia here you go:
```
[root@kube-7 ~]# kubectl get pods -n istio-system istio-pilot-755579f496-rcv4d -o yaml
apiVersion: v1
kind: Pod
metadata:
  annotations:
    scheduler.alpha.kubernetes.io/critical-pod: ""
    sidecar.istio.io/inject: "false"
  creationTimestamp: "2019-03-06T21:50:37Z"
  generateName: istio-pilot-755579f496-
  labels:
    app: pilot
    chart: pilot
    heritage: Tiller
    istio: pilot
    pod-template-hash: 755579f496
    release: istio
  name: istio-pilot-755579f496-rcv4d
  namespace: istio-system
  ownerReferences:
  - apiVersion: apps/v1
    blockOwnerDeletion: true
    controller: true
    kind: ReplicaSet
    name: istio-pilot-755579f496
    uid: e049840e-4059-11e9-8816-525400cd2379
  resourceVersion: "459674"
  selfLink: /api/v1/namespaces/istio-system/pods/istio-pilot-755579f496-rcv4d
  uid: e04ac7e4-4059-11e9-8816-525400cd2379
spec:
  affinity:
    nodeAffinity:
      preferredDuringSchedulingIgnoredDuringExecution:
      - preference:
          matchExpressions:
          - key: beta.kubernetes.io/arch
            operator: In
            values:
            - amd64
        weight: 2
      - preference:
          matchExpressions:
          - key: beta.kubernetes.io/arch
            operator: In
            values:
            - ppc64le
        weight: 2
      - preference:
          matchExpressions:
          - key: beta.kubernetes.io/arch
            operator: In
            values:
            - s390x
        weight: 2
      requiredDuringSchedulingIgnoredDuringExecution:
        nodeSelectorTerms:
        - matchExpressions:
          - key: beta.kubernetes.io/arch
            operator: In
            values:
            - amd64
            - ppc64le
            - s390x
  containers:
  - args:
    - discovery
    - --monitoringAddr=:15014
    - --domain
    - k7.sbezverk.cisco.com
    - --trust-domain
    - k7.sbezverk.cisco.com
    - --secureGrpcAddr
    - ""
    - --keepaliveMaxServerConnectionAge
    - 30m
    env:
    - name: POD_NAME
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: metadata.name
    - name: POD_NAMESPACE
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: metadata.namespace
    - name: GODEBUG
      value: gctrace=2
    - name: PILOT_PUSH_THROTTLE_COUNT
      value: "100"
    - name: PILOT_TRACE_SAMPLING
      value: "1"
    image: 192.168.80.240:4000/istio/pilot:v1.2.1_ip6tablesv6_local_cluster
    imagePullPolicy: Always
    name: discovery
    ports:
    - containerPort: 8080
      protocol: TCP
    - containerPort: 15010
      protocol: TCP
    readinessProbe:
      failureThreshold: 3
      httpGet:
        path: /ready
        port: 8080
        scheme: HTTP
      initialDelaySeconds: 5
      periodSeconds: 30
      successThreshold: 1
      timeoutSeconds: 5
    resources:
      requests:
        cpu: 500m
        memory: 2Gi
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /etc/istio/config
      name: config-volume
    - mountPath: /etc/certs
      name: istio-certs
      readOnly: true
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: istio-pilot-service-account-token-zdvg5
      readOnly: true
  - args:
    - proxy
    - --domain
    - $(POD_NAMESPACE).svc.k7.sbezverk.cisco.com
    - --trust-domain
    - sbezverk.cisco.com
    - --serviceCluster
    - istio-pilot
    - --templateFile
    - /etc/istio/proxy/envoy_pilot.yaml.tmpl
    - --controlPlaneAuthPolicy
    - NONE
    env:
    - name: POD_NAME
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: metadata.name
    - name: POD_NAMESPACE
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: metadata.namespace
    - name: INSTANCE_IP
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: status.podIP
    image: 192.168.80.240:4000/istio/proxyv2:v1.2.1_ip6tablesv6_local_cluster
    imagePullPolicy: Always
    name: istio-proxy
    ports:
    - containerPort: 15003
      protocol: TCP
    - containerPort: 15005
      protocol: TCP
    - containerPort: 15007
      protocol: TCP
    - containerPort: 15011
      protocol: TCP
    resources:
      limits:
        cpu: "2"
        memory: 128Mi
      requests:
        cpu: 100m
        memory: 128Mi
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /etc/certs
      name: istio-certs
      readOnly: true
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: istio-pilot-service-account-token-zdvg5
      readOnly: true
  dnsPolicy: ClusterFirst
  enableServiceLinks: true
  nodeName: kube-7.sbezverk.cisco.com
  priority: 0
  restartPolicy: Always
  schedulerName: default-scheduler
  securityContext: {}
  serviceAccount: istio-pilot-service-account
  serviceAccountName: istio-pilot-service-account
  terminationGracePeriodSeconds: 30
  tolerations:
  - effect: NoExecute
    key: node.kubernetes.io/not-ready
    operator: Exists
    tolerationSeconds: 300
  - effect: NoExecute
    key: node.kubernetes.io/unreachable
    operator: Exists
    tolerationSeconds: 300
  volumes:
  - configMap:
      defaultMode: 420
      name: istio
    name: config-volume
  - name: istio-certs
    secret:
      defaultMode: 420
      optional: true
      secretName: istio.istio-pilot-service-account
  - name: istio-pilot-service-account-token-zdvg5
    secret:
      defaultMode: 420
      secretName: istio-pilot-service-account-token-zdvg5
status:
  conditions:
  - lastProbeTime: null
    lastTransitionTime: "2019-03-06T21:50:37Z"
    status: "True"
    type: Initialized
  - lastProbeTime: null
    lastTransitionTime: "2019-03-06T22:24:31Z"
    message: 'containers with unready status: [discovery]'
    reason: ContainersNotReady
    status: "False"
    type: Ready
  - lastProbeTime: null
    lastTransitionTime: "2019-03-06T22:24:31Z"
    message: 'containers with unready status: [discovery]'
    reason: ContainersNotReady
    status: "False"
    type: ContainersReady
  - lastProbeTime: null
    lastTransitionTime: "2019-03-06T21:50:37Z"
    status: "True"
    type: PodScheduled
  containerStatuses:
  - containerID: docker://5bf6774126802569fe1bec34fc3de454e1c07e5c51e06d04d541c4af3150b03c
    image: 192.168.80.240:4000/istio/pilot:v1.2.1_ip6tablesv6_local_cluster
    imageID: docker-pullable://192.168.80.240:4000/istio/pilot@sha256:2ef1ec04e62c3f6282125ea39a8948f6c722864885b24ba39ccd72cc9119a058
    lastState: {}
    name: discovery
    ready: false
    restartCount: 0
    state:
      running:
        startedAt: "2019-03-06T21:50:44Z"
  - containerID: docker://5b0fc2c2510af723bdb6d86e79bd8a29c6b05cc06799d14ff38251483c2ed51f
    image: 192.168.80.240:4000/istio/proxyv2:v1.2.1_ip6tablesv6_local_cluster
    imageID: docker-pullable://192.168.80.240:4000/istio/proxyv2@sha256:071351a550bf837344b8ded30c66be90b36153ec38c48f80486a63bb19c5a525
    lastState: {}
    name: istio-proxy
    ready: true
    restartCount: 0
    state:
      running:
        startedAt: "2019-03-06T21:50:47Z"
  hostIP: 2001:470:b16e:81::11
  phase: Running
  podIP: 2001:470:b16e:1001::5678
  qosClass: Burstable
  startTime: "2019-03-06T21:50:37Z"
[root@kube-7 ~]# 
```

**Comment by andraxylia** [comment]
Thanks, it does say "message: 'containers with unready status: [discovery]'

**Comment by andraxylia** [comment]
Turns out this is a pure IPv6 cluster and the issue is not with the pilot, kubelet cannot access the readiness probe anymore.

Could be with ipv6 support or kubelt bug for ipv6.

**Comment by Miouge1** [comment]
I'm seeing similar errors in my 1.1.0 test deployment IPv4:

```
NAME                                      READY   STATUS             RESTARTS   AGE
istio-citadel-8c574d7c8-bwshw             1/1     Running            0          69m
istio-galley-9d5986bc8-ffmzm              1/1     Running            0          69m
istio-ingressgateway-5b6d8f9ff8-nffn7     0/1     Running            0          69m
istio-pilot-6485f4b679-hs4pn              1/2     Running            0          67m
istio-policy-77f645d6f-v542h              1/2     CrashLoopBackOff   16         69m
istio-sidecar-injector-57d48d8f6f-bbbtk   1/1     Running            0          69m
istio-telemetry-b9cc9c85-jm92q            1/2     CrashLoopBackOff   27         69m
prometheus-8647cf4bc7-c4cpk               1/1     Running            0          69m
```

And lots of `Configuration not synced: first push for ...` in the pilot discovery logs.

Any idea where to dig?

**Comment by ozevren** [comment]
@Miouge1 can you share galley & mixer logs?
kubectl logs -n istio-system istio-galley-9d5986bc8-ffmzm 
kubectl logs -n istio-system istio-policy-77f645d6f-v542h

**Comment by Miouge1** [comment]
@ozevren sure:
- [Istio Galley logs](http://paste.openstack.org/show/3JwTzVB4NFPensLHYuOG/)
- [Istio Policy mixer logs](http://paste.openstack.org/show/wl6RUwLpl2MZ53mttqf4/)

**Comment by ozevren** [comment]
It looks like Mixer (and possibly Pilot) is not able to connect to Galley. There are no incoming connection entry logs  for Galley. Mixer logs is showing attempts for connection:

```
2019-03-26T15:42:26.770894Z	info	ccResolverWrapper: sending new addresses to cc: [{istio-galley.istio-system.svc:9901 0  <nil>}]
2019-03-26T15:42:26.770951Z	info	ClientConn switching balancer to "pick_first"
2019-03-26T15:42:26.771174Z	info	pickfirstBalancer: HandleSubConnStateChange: 0xc4205220e0, CONNECTING
2019-03-26T15:42:26.771006Z	info	mcp	(re)trying to establish new MCP sink stream
2019-03-26T15:42:46.771594Z	info	grpc: addrConn.createTransport failed to connect to {istio-galley.istio-system.svc:9901 0  <nil>}. Err :connection error: desc = "transport: Error while dialing dial tcp: i/o timeout". Reconnecting...
2019-03-26T15:42:46.771700Z	info	pickfirstBalancer: HandleSubConnStateChange: 0xc4205220e0, TRANSIENT_FAILURE
```

**Comment by Miouge1** [comment]
That's right. Access to Galley's TCP/443 works locally:
```
root@istio-galley-9d5986bc8-ffmzm:/# curl -k https://localhost/
404 page not found
```
But doesn't work from any of the kube-node. Any ideas?

**Comment by zqzzq** [comment]
I have the same problem in 1.1.3, the discovery in pilot becomes Unhealthy constantly:
```
Events:
  Type     Reason     Age                 From              Message
  ----     ------     ----                ----              -------
  Warning  Unhealthy  4m (x2541 over 1d)  kubelet, master3  Readiness probe failed: Get http://158.158.72.16:8080/ready: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
 ```

but when I curl http://158.158.72.16:8080/ready ,I get 200 :
``` 
kubectl -n istio-system exec -it istio-pilot-8647cf5859-bbzwf -- curl -iv http://158.158.72.16:8080/ready
Defaulting container name to discovery.
Use 'kubectl describe pod/istio-pilot-8647cf5859-bbzwf -n istio-system' to see all of the containers in this pod.
*   Trying 158.158.72.16...
* Connected to 158.158.72.16 (158.158.72.16) port 8080 (#0)
> GET /ready HTTP/1.1
> Host: 158.158.72.16:8080
> User-Agent: curl/7.47.0
> Accept: */*
> 
< HTTP/1.1 200 OK
HTTP/1.1 200 OK
< Date: Sat, 20 Apr 2019 07:29:51 GMT
Date: Sat, 20 Apr 2019 07:29:51 GMT
< Content-Length: 0
Content-Length: 0

< 
* Connection #0 to host 158.158.72.16 left intact 
```
and in pilot discovery logs I found many error like this：
``` 
2019-04-20T07:39:51.746955Z     info    transport: http2Server.HandleStreams failed to read frame: read tcp 158.158.72.16:15010->158.158.111.159:33250: use of closed network connection
2019-04-20T07:39:52.346730Z     info    transport: loopyWriter.run returning. connection error: desc = "transport is closing"
2019-04-20T07:39:52.646864Z     info    transport: loopyWriter.run returning. connection error: desc = "transport is closing"
2019-04-20T07:39:52.746730Z     info    transport: http2Server.HandleStreams failed to read frame: read tcp 158.158.72.16:15010->158.158.111.132:46410: use of closed network connection
2019-04-20T07:39:52.248945Z     error   ads     ADS: "158.158.111.132:46410" sidecar~158.158.111.132~app-91qau-6b55955944-6dnxf.default~default.svc.cluster.local-205 terminated with stream closed
  ```

**Comment by Miouge1** [comment]
In my case it was caused by a misconfigured CNI. Once I fixed the CNI config, istio was happy.

**Comment by gobomb** [comment]
> In my case it was caused by a misconfigured CNI. Once I fixed the CNI config, istio was happy.

how to fix?

**Comment by xuemzhan** [comment]
> misconfigured

I met the same problem, how to fix it?

**Comment by Miouge1** [comment]
In my cluster I use Calico for CNI and I had IP-in-IP encapsulation enabled by accident. Therefore pod to pod communication was broken to start with, even without Istio installed.

**Comment by dghant1024** [comment]
@Miouge1 Thanks for the insight.  I'm running into the same identical issue ... I'm curious to know if you changed the CNI to use a different CNI, or if are you still using Calico?  If I recall correctly, Calico natively uses flannel to deliver package payloads "on the wire." I also understand that Calico supports other networking protocols (ex:BGP).  Did you change to use a different CNI (ex: Weave, romana, etc)?  Last question, other then curl-ing urls from within the container, did use any other tools to identify that it was a networking issue (tcpdump, ipvsadm)?  If other are willing to give insights on troubleshooting things like this I'm listening ... lol.

**Comment by Miouge1** [comment]
@dghant1024 Calico has an IP-in-IP mode, which I do not use (and I think it's different than Flannel's, you might be mixing up with Canal). Instead I use the routing mode which delivers wire speeds.

`curl` was enough to findout that something was wrong with the CNI.

Good lock.

**Comment by dghant1024** [comment]
Thanks @Miouge1 You're correct, I'm asking because I use Canal which offers Calico.  I may be wrong, but the way I understand Canal is that it maximizes the policy features offered by Calico and allows the use different network typologies (BGP or IP-in-IP) for delivery.  But, I think if neither protocol is specified for Calico, as Canal configuration, (I think) Canal defaults to flannel to deliver its payload !!! But I'm not absolutely sure.  Too may layers to parse through when dealing with default setting of Canal. I looks like I may need to see how settings are defaulting in my configuration of Canal and make sure that IP-in-IP is not deployed to the cluster.   Thanks again.

**Comment by sdake** [comment]
I am unclear as to the root cause however if Pilot is not able to connect to galley, it should not cause Pilot to enter a loop of failure.

**Comment by sdake** [comment]
https://github.com/istio/istio.io/issues/4121

**Comment by howardjohn** [comment]
@sbezverk sounds like root cause was incomplete ipv6 support. Can this be closed now that we have ipv6 support?

**Comment by sbezverk** [comment]
It seems it happens from time to time even on ipv4 installations, but since it is not reliably reproducible, I think it can be closed.

**Comment by howardjohn** [comment]
Ok I think we can close this then, otherwise tickets like this tend to become a collector of +1s that are hard to track and understand because everyone's issues have different causes . If you do run into it again we can open a new issue and track there
